File: week_3/quicksort/quicksort.py
from random import shuffle
import unittest

def quicksort(lst):
    return (quicksort([x for x in lst[1:] if x < lst[0]]) +
           lst[:1] + 
           quicksort([x for x in lst[1:] if x >= lst[0]])) if len(lst) > 1 else lst

class QuickSortTest(unittest.TestCase):
  
    def setUp(self):
        self.ary = list(range(10))
        shuffle(self.ary)
 
    def test_initialize(self):
        self.assertEqual(len(self.ary), 10)
        self.assertNotEqual(self.ary, list(range(10)))

    def test_quicksort(self):
        self.assertEqual(quicksort(self.ary), list(range(10)))

if __name__ == '__main__':
    unittest.main()

File: week_3/quicksort/quicksort.rb
require 'test/unit'

class Array

  def quick_sort
    h, *t = self
    h ? t.partition { |e| e < h }.inject { |l, r| l.quick_sort + [h] + r.quick_sort } : []
  end

end


class QuickSort < Test::Unit::TestCase

  def setup
    @ary = (0..9).to_a.shuffle
    
  end

  def test_initialize
    assert_equal(@ary.size, 10)
    assert_not_equal(@ary.size, (0..9).to_a)
  end

  def test_quick_sort
    assert_equal(@ary.quick_sort, (0..9).to_a)
  end

end